packages/babel-traverse/src/path/inference/inferer-reference.js

```diff
@@ -91,9 +91,15 @@ function getTypeAnnotationBindingConstantViolations(binding, path, name) {
     }
   }
 
-  if (types.length) {
-    return t.createUnionTypeAnnotation(types);
+  if (!types.length) {
+    return;
+  }
+
+  if (t.isTSTypeAnnotation(types[0])) {
+    return t.createTSUnionType(types);
```

Comment:
This is technically a breaking change: it makes the next `@babel/traverse` version incompatible with `@babel/types` <= 7.9.5, because it doesn't contain this function.

We should use `t.createTSUnionType?.(types)` instead.

Reply:
Oh, this is true, thanks for mentioning that

Reply:
@nicolo-ribaudo it will change the behaviour as returns `UnionTypeAnnotation` and now it returns `undefined`

I added if-check in https://github.com/babel/babel/pull/11378/commits/5e9c92098d7d79a9e6f0c72d67370d0b2400c9cf

✅